crypto/encode_decode/decoder_pkey.c

```diff
@@ -438,7 +443,9 @@ static int ossl_decoder_ctx_setup_for_pkey(OSSL_DECODER_CTX *ctx,
     collect_data.libctx     = libctx;
     collect_data.keymgmts   = keymgmts;
     collect_data.keytype    = keytype;
-    EVP_KEYMGMT_do_all_provided(libctx, collect_keymgmt, &collect_data);
+    EVP_KEYMGMT_do_all_provided(libctx,
+                                (void (*)(EVP_KEYMGMT *, void*))collect_keymgmt_thunk,
```

Comment:
Needing a cast here is pretty odd. If `EVP_KEYMGMT_do_all_provided` actually wants an untyped function pointer, it should just be declared as such. If it wants a typed function pointer, its implementation should reflect this.

Reply:
This particular cast wasnt/isn't about thunking per-se, its the result of a compiler warning (promoted to an error since I'm building with strict-warnings).

collecT_keymgmt_thunk is typed as:
```
static void collect_keymgmt_thunk(void *data, void *arg)
```
but EVP_KEYMGMT_do_all_provided accepts the following:
```
void EVP_KEYMGMT_do_all_provided(OSSL_LIB_CTX *libctx,
                                 void (*fn)(EVP_KEYMGMT *keymgmt, void *arg),
                                 void *arg)
```
Because of the:
```
void (*fn)(EVP_KEYMGMT *keymgmt, void *arg)
```
The compiler warns about mismatched types in the parameter list

I could of course change EVP_KEYMGMT_do_all_provided to accept void(*fn)(void *km, void *arg), but given that EVP_KEYMGMT_do_all_provided is a public facing abi, I thought this was the better approach

Reply:
@nhorman Instead of adding a thunk, modify the filter_on_operation_id() in evp_fetch.c to have a switch() based on the id, that does the cast to the concrete function type before the call:
```
static void filter_on_operation_id(int id, void *method, void *arg)
{
    struct filter_data_st *data = arg;

    if ((id & METHOD_ID_OPERATION_MASK) == data->operation_id)
        switch (data->operation_id) {
        case OSSL_OP_KEYMGMT:
            (void (*)(EVP_KEYMGMT *, void *))data->user_fn((EVP_KEYMGMT *)method, data->user_arg);
        case ...
        }
}
```

Reply:
I could do that, but I'm not sure how I see that as adventageous.  It swaps in a potentially large case statement in place of a cast at the location where we know what the typing was.

Reply:
Hmm... but then all the callers of EVP_KEYMGMT_do_all_provided() will have to change their calls to avoid undefined behavior anyway. IMO this is wrong approach. You could also do the thunking inside EVP_KEYMGMT_do_all_provided().

Reply:
@nhorman I would suggest stopping this work temporarily and leaving this for OTC discussion on how we want to handle this. As one option could clearly be to keep the code as is until for example 4.0 where we can introduce API incompatibilities.

Reply:
Thats what I suggested above, in altering the function signature that we pass as a parameter to be
```
 void(*fn)(void *, void *)
```
and thats certainly workable, but it changes the api (which again is ok, was just trying to avoid it

The one thing we can't do is change the prototype of collect_keymgmt_thunk, as that re-instates the ubsan warning when the call back is made.

The options here that I see are:
1) change the signature of EVP_KEYMGMT_do_all_provided to accept the above function type
or
2) Do your thunking approach in filter_on_operation_id

Option 1 seems reasonable if we're ok with it.  Option 2 results in an (ostensibly large) jump table that will have to be done for the user_fn callback, and the free method callback.

Reply:
There is an option 3 - the thunk can be done in EVP_KEYMGMT_do_all_provided() - of course the thunk will have to be more complicated as it will have to pass a wrapper data structure pointer as the arg that would contain the original function call and the original data arg instead of just the original data arg.

Reply:
Thats...thats what collect_keymgmt_thunk is, a wrapper that is passed to do the thunking.

This change was done because I hit this warning when compiling:
```
crypto/encode_decode/decoder_pkey.c:447:33: error: incompatible function pointer types passing 'void (void *, void *)' to parameter of type 'void (*)(EVP_KEYMGMT *, void *)' (aka 'void (*)(struct evp_keymgmt_st *, void *)') [-Wincompatible-function-pointer-types]
  447 |                                 collect_keymgmt_thunk,
      |                                 ^~~~~~~~~~~~~~~~~~~~~
include/openssl/evp.h:1775:41: note: passing argument to parameter 'fn' here
 1775 |                                  void (*fn)(EVP_KEYMGMT *keymgmt, void *arg),
```

There are 2 ways to fix that:
1) Cast collect_keymgmt_thunk to void(*)(EVP_KEYMGMT *, void *)
2) Change the type of collect_keymgmt_thunk to be void collect_keymgmt_thunk(EVP_KEYMGMT *k, void *a)

Option 1 is what you see in the code now.

Option 2 eliminates the above error, but re-introduces a ubsan error at run time, when filter_on_operation_id calls data->user_fn, because the user_fn type of void (*)(void *, void *) doesn't match the target of the pointer, which would be void (*)(EVP_KEYMGMT *, void *).

As you noted previously, we could avoid that by creating  a jump table in filter_on_operation_id, to cast user_fn to the appropriate type for the operation, but that would require a (currently) 13 case case statement that would need to be augmented for any new operation type.

I think its the wiser move to create a thunking function for each operation to do the cast, as that allows us to co-locate the thunking code with the actual operational callback, which is more intuitive in my mind (i.e. you will see the thunking code in close spatial proximity to the actual operation), rather than having to remember that the casting happens else where, and needs to be updated if you add a new operation type